File: Runtime/TransitionOrderMap.vb
'*********************************************************************************************
' ST-Sim: A SyncroSim Module for the ST-Sim State-and-Transition Model.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'*********************************************************************************************

Imports SyncroSim.Common

Class TransitionOrderMap

    Private m_HasItems As Boolean
    Private m_Map As New SortedKeyMap2(Of TransitionOrderCollection)(SearchMode.ExactPrev)

    Public Sub New(ByVal orders As TransitionOrderCollection)

        For Each t As TransitionOrder In orders
            Me.AddTransitionOrder(t)
        Next

    End Sub

    Private Sub AddTransitionOrder(ByVal order As TransitionOrder)

        Dim l As TransitionOrderCollection = Me.m_Map.GetItemExact(order.Iteration, order.Timestep)

        If (l Is Nothing) Then

            l = New TransitionOrderCollection()
            Me.m_Map.AddItem(order.Iteration, order.Timestep, l)

        End If

        l.Add(order)

        Me.m_HasItems = True

    End Sub

    Public Function GetTransitionOrders(ByVal iteration As Integer, ByVal timestep As Integer) As TransitionOrderCollection

        If (Not Me.m_HasItems) Then
            Return Nothing
        End If

        Dim l As TransitionOrderCollection = Me.m_Map.GetItem(iteration, timestep)

        If (l Is Nothing) Then
            Return Nothing
        End If

        Debug.Assert(l.Count > 0)
        Return l

    End Function

End Class

File: src/Runtime/TransitionOrderMap.cs
// A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2019 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.

using SyncroSim.Common;
using System.Diagnostics;

namespace SyncroSim.STSim
{
    internal class TransitionOrderMap
    {
        private bool m_HasItems;
        private SortedKeyMap2<TransitionOrderCollection> m_Map = new SortedKeyMap2<TransitionOrderCollection>(SearchMode.ExactPrev);

        public TransitionOrderMap(TransitionOrderCollection orders)
        {
            foreach (TransitionOrder t in orders)
            {
                this.AddTransitionOrder(t);
            }
        }

        private void AddTransitionOrder(TransitionOrder order)
        {
            TransitionOrderCollection l = this.m_Map.GetItemExact(order.Iteration, order.Timestep);

            if (l == null)
            {
                l = new TransitionOrderCollection();
                this.m_Map.AddItem(order.Iteration, order.Timestep, l);
            }

            l.Add(order);

            this.m_HasItems = true;
        }

        public TransitionOrderCollection GetTransitionOrders(int iteration, int timestep)
        {
            if (!this.m_HasItems)
            {
                return null;
            }

            TransitionOrderCollection l = this.m_Map.GetItem(iteration, timestep);

            if (l == null)
            {
                return null;
            }

            Debug.Assert(l.Count > 0);
            return l;
        }
    }
}
